### DISCUSSION TITLE
Introduce automate tests and choose the test framework

### DISCUSSION THREAD

**Discussion by tbouffard** [maxGraph/maxGraph#153] [Development]
Hi all,
mxGraph did not have automatic tests so it only relied on manual testing and new releases often introduced regressions.

For now, maxGraph does not have them either. We have done the full migration and changed some parts of the implementation without taking the opportunity to add some.
This is a subject we mentioned several times, for instance in April 2021 (#18) and on 2021-12-14 (https://github.com/maxGraph/maxGraph/discussions/64#discussioncomment-1808782). But we left it as it was.

So today we have problems and regressions compared to the old implementation of mxGraph.
I have just fixed a style issue in #152 because of that. As part of the fix, I have started to introduce tests and I would like us to generalize automatic tests.

I suggest we start adding automatic low-level/micro/unit tests (name them as you want) when fixing bugs and implementing new features. This will gradually increase the code coverage at a sustainable rate.
Utility functions can be easily tested by such tests, and it should also be feasible to test the code involving DOM manipulations.
maxGraph also involves interactions and rendering. We could investigate the best ways to test these parts too.

### Experimentation

In #152 and #159, I used `jest` as a test framework and `ts-jest` for type check support. `jest` is widely used, actively maintained and a lot of community extensions exist.
The `ts-jest` speed will increase in the future as it will use esbuild/swc for transpilation. Currently, generation takes several seconds prior running the tests.
As maxGraph uses a lot of browser objects, also setup `jest-jsdom-environment`.

#152 and #159 use the following mini-conventions:
- store tests in the `__tests__` directory (jest default)
- file name: <filename_under_test>.test.ts
- use describe and test, standard jest expect (no extensions)

This is just a starting point. We can decide to use another framework and other conventions.
Test frameworks use similar concepts and API, so migrating to another one will be straightforward.

### My Proposal

I would like us to move quickly on this subject, by starting with unit tests:
- don't wait for the test framework choice and use jest for now in #152 to not block the merge of the fix.
- choose a framework and conventions
- migrate or adjust existing tests
- document that adding tests is required (in PR template?)
- enforce this rule (automatically or during review)

Then, we could continue with tests about rendering and interactions.

Feel free to comment.
cc @aminosbh @cd-yang @csouchet @junsikshim @mayorovad @mcyph @zhangbihua

**Comment by junsikshim** [comment]
I agree that it's time to add tests to maxgraph. 
I don't usually test UI as it changes way too often, but since this is a library, it makes sense to have them.
Using `jest`, `ts-jest`, and `jest-environment-jsdom` is definitely a fine choice to begin with.

We could perhaps get new contributors making only the unit tests before they get accustomed to the codebase.

**Reply by tbouffard** [reply]
Thanks for your feedback. #156 has already been merged, so everybody can/should/must 😸 add tests.

> We could perhaps get new contributors making only the unit tests before they get accustomed to the codebase.

Good idea, I will create an issue and pin it to pass the message.

About visual/UI testing, I agree with you about the fact they can change often and so can be hard to maintain. I have something in mind, based on what we learn when developing the `bpmn-visualization` library. Here is some elements why we are using them in this library: https://github.com/process-analytics/bpmn-visualization-js/blob/v0.29.0/docs/contributors/testing.md#visual-testing. In particular, they caught a lot of regressions/bugs in `maxGraph` while experimenting the migration from `mxGraph` and they help me to submit fix or issues. 

That being said, I think this topic needs a specific decisions about what we want to test in `maxGraph` and why. Then, if we go on that way, we will have to choose the right tool. I have some thoughts about that that I will summarize in a dedicated discussion later.
